src/main/scala/scripts/s3wrapper/S3Wrapper.scala

```diff
@@ -0,0 +1,182 @@
+    cachedBuckets.get()
+  }
+
+  def refreshBucketNames: Set[S3Bucket] = {
+    cachedBuckets.updateAndGet(_ => listBucketNamesInternal)
+  }
+
+  private def listBucketNamesInternal: Set[S3Bucket] = {
+    s3Client
+      .listBuckets()
+      .asScala
+      .par
+      .map { b =>
+        val location = s3Client.getBucketLocation(b.getName)
+        S3Bucket(b.getName, location)
+      }
+      .toSet
+      .seq
+  }
+
+  /**
+    * Get recursively all keys belongs to some bucket. Since AWS API limits to 1000
+    * the maximum amount of keys per request, it might be needed to apply multiple requests
+    * The CLI equivalent command is as following:
+    * `aws s3 ls --summarize --recursive s3://beam-admin-reports/ --profile beam`
+    *
+    * @param bucketName  the name of the bucket
+    * @param maxRequests the maximum number of requests. Default value is unlimited in practical terms (Int.MaxValue)
+    * @return a set with all keys from the bucket
+    */
+  def allBucketFiles(bucketName: String, maxRequests: Int = Int.MaxValue): Set[S3RemoteFile] = {
+    val request = new ListObjectsV2Request()
+      .withBucketName(bucketName)
+      .withMaxKeys(1000)
+    allBucketFiles(request, Set.empty, maxRequests)
+  }
+
+  @tailrec
+  private def allBucketFiles(request: ListObjectsV2Request, acc: Set[S3RemoteFile], max: Int): Set[S3RemoteFile] = {
+    if (max > 0) {
+      val result: ListObjectsV2Result = s3Client.listObjectsV2(request)
+      val resultKeys = result
+        .getObjectSummaries
+        .asScala
+        .map(o => S3RemoteFile(o.getKey, o.getSize))
+        .toSet
+      if (result.isTruncated) {
+        val token = result.getNextContinuationToken
+        val newRequest = request.withContinuationToken(token)
+        allBucketFiles(newRequest, acc ++ resultKeys, max - 1)
+      } else {
+        acc ++ resultKeys
+      }
+    } else {
+      acc
+    }
+  }
+
+  /**
+    * Download the full bucket and save into the destination path
+    * @param bucketName the bucket name
+    * @param destinationPath the destination path
+    */
+  def downloadBucket(bucketName: String, destinationPath: Path): Unit = {
+    download(bucketName, wholeDirectory, destinationPath)
+  }
+
+  /**
+    * Download all files, directories and subdirectories accordingly to the parameters
+    * @param bucketName the bucket name
+    * @param keysStartingWith the object key is composed its full path
+    * @param destinationPath the directory which downloaded data will be saved
+    */
+  final def download(bucketName: String, keysStartingWith: String, destinationPath: Path): Unit = {
+    if (!Files.isDirectory(destinationPath)) {
+      throw new IllegalArgumentException(s"Destination path [$destinationPath] must be a directory")
+    }
+    if (Files.exists(destinationPath.resolve(bucketName))) {
+      throw new IllegalArgumentException(s"Destination path [$destinationPath] cannot contain a directory with the bucket name [$bucketName]")
```

Comment:
I propose different text for exception: 
Destination path [$destinationPath] already contains a directory with the name [$bucketName]''